src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -270,82 +476,109 @@ impl Coordinator {
 
         let transact_result = self
             .catalog_transact_with_side_effects(Some(session), ops, |coord| async {
-                let mut read_policy_updates = Vec::new();
+                // Buffer an ingestions' read policies in case their subsources
+                // need to reference it before it's durably recorded.
+                let mut ingestion_read_policies: BTreeMap<_, _> = BTreeMap::new();
+                let mut read_policies: BTreeMap<Option<CompactionWindow>, Vec<GlobalId>> =
+                    BTreeMap::new();
 
-                for (source_id, source) in sources {
+                let mut collections = Vec::with_capacity(sources.len());
+
+                for (source_id, mut source) in sources {
                     let source_status_collection_id =
                         Some(coord.catalog().resolve_builtin_storage_collection(
                             &mz_catalog::builtin::MZ_SOURCE_STATUS_HISTORY,
                         ));
 
-                    let (data_source, status_collection_id, set_read_policies) = match source
-                        .data_source
-                    {
+                    let (data_source, status_collection_id) = match source.data_source {
                         DataSourceDesc::Ingestion(ingestion) => {
                             let ingestion =
                                 ingestion.into_inline_connection(coord.catalog().state());
 
-                            // The parent source dictates all of the subsource's read policies.
-                            let set_read_policies =
-                                ingestion.source_exports.keys().cloned().collect();
+                            ingestion_read_policies
+                                .insert(source_id, source.custom_logical_compaction_window);
 
                             (
                                 DataSource::Ingestion(ingestion),
                                 source_status_collection_id,
-                                set_read_policies,
+                            )
+                        }
+                        DataSourceDesc::SourceExport {
+                            id,
+                            external_reference,
+                        } => {
+                            // Propagate source's compaction window if this
+                            // subsource does not have its own value specified.
```

Comment:
We re-set these during bootstrapping, but it doesn't look like the same logic gets applied. cc @mjibson

Reply:
cc @maddyblue -- any idea what we ought to do here, or have you moved onto other projects?

Reply:
Reading over https://github.com/MaterializeInc/materialize/pull/24010/files it looks like on restart subsource retain history does not inherit from the source. #26650 was already tracking another cause of this (so subsources were doubly broken here). I think this PR should not worry about that because it's already big enough and doesn't work anyway. I'll fix it later.

Reply:
Thank you!